### PR TITLE
docs: add fip-4, bump to 2023.7.12

### PR DESCRIPTION
**Pull request by varunsrin**

### CODE REVIEW

File: docs/SPECIFICATION.md [hunk 1]
```diff
@@ -424,6 +427,49 @@ A LinkRemove in a message `m` is valid only if it passes these validations:
 
 1.  `m.data.type` must be `MESSAGE_TYPE_LINK_REMOVE`
 
+## 1.7 Username Proof
+
+```protobuf
+enum UserNameType {
+  USERNAME_TYPE_NONE = 0;
+  USERNAME_TYPE_ENS_FNAME = 1;
```
Comment (sanjayprabhu) [inline docs/SPECIFICATION.md:435]: This is just called `USERNAME_TYPE_FNAME`

File: docs/SPECIFICATION.md [hunk 2]
```diff
@@ -424,6 +427,49 @@ A LinkRemove in a message `m` is valid only if it passes these validations:
 
 1.  `m.data.type` must be `MESSAGE_TYPE_LINK_REMOVE`
 
+## 1.7 Username Proof
```
Comment (sanjayprabhu) [inline docs/SPECIFICATION.md:430]: `USER_DATA_TYPE_FNAME` has been renamed to `USER_DATA_TYPE_USERNAME` in the UserDataAdd spec above.

File: docs/SPECIFICATION.md [hunk 3]
```diff
@@ -424,6 +427,49 @@ A LinkRemove in a message `m` is valid only if it passes these validations:
...
+3. `m.data.body.signature` must be a valid ECDSA signature on the EIP-712 Username Proof message from the owner or the public key of the fname server.
+
+A UsernameProof message `m` of type `USERNAME_TYPE_ENS_L1` must also pass these validations:
+
+1. `m.data.body.name` name must:
+   1. be a valid, unexpired ENS name
+   2. match the regular expression `/^[a-z0-9][a-z0-9-]{0,15}$/`
+   3. end in `.eth`
+2. `m.data.body.owner` must:
+   1. be an address that the user has a valid VerificationMessage for.
```
Comment (sanjayprabhu) [inline docs/SPECIFICATION.md:469]: It can also be the custody address

File: docs/SPECIFICATION.md [hunk 4]
```diff
@@ -424,6 +427,49 @@ A LinkRemove in a message `m` is valid only if it passes these validations:
...
+   2. be the owner of the ENS name.
```
Comment (sanjayprabhu) [inline docs/SPECIFICATION.md:470]: ENS name must resolve to the `owner` is more accurate. The actual registrant or controller can be different.

File: docs/SPECIFICATION.md [hunk 5]
```diff
@@ -424,6 +427,49 @@ A LinkRemove in a message `m` is valid only if it passes these validations:
...
+
+1. `m.data.body.name` name must match the regular expression `/^[a-z0-9][a-z0-9-]{0,15}$/`.
+2. `m.data.body.owner` must be the custody address of the fid.
+3. `m.data.body.signature` must be a valid ECDSA signature on the EIP-712 Username Proof message from the owner or the public key of the fname server.
+
+A UsernameProof message `m` of type `USERNAME_TYPE_ENS_L1` must also pass these validations:
+
+1. `m.data.body.name` name must:
+   1. be a valid, unexpired ENS name
+   2. match the regular expression `/^[a-z0-9][a-z0-9-]{0,15}$/`
```
Comment (sanjayprabhu) [inline docs/SPECIFICATION.md:466]: This only applies to the non `.eth` portion. So, actual regex is more like `^[a-z0-9][a-z0-9-]{0,15}\.eth$`

File: docs/SPECIFICATION.md (hunk 1 above)
Comment (varunsrin) [reply docs/SPECIFICATION.md:435]: should we rename this to ENS_FNAME in the implementation? 

the rationale is that this makes it clear that the fname is a type of ENS name, not something distinct

Comment (sanjayprabhu) [reply docs/SPECIFICATION.md:435]: Sure, will rename.

**Pull request merged**